pqarrow/arrowutils/sort.go

```diff
@@ -2,93 +2,306 @@ package arrowutils
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"sort"
 
 	"github.com/apache/arrow/go/v14/arrow"
 	"github.com/apache/arrow/go/v14/arrow/array"
 	"github.com/apache/arrow/go/v14/arrow/compute"
 	"github.com/apache/arrow/go/v14/arrow/memory"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/polarsignals/frostdb/pqarrow/builder"
+)
+
+type Direction uint
+
+const (
+	Ascending Direction = iota
+	Descending
 )
 
-// SortRecord sorts the given record's rows by the given column. Currently only supports int64, string and binary columns.
-func SortRecord(mem memory.Allocator, r arrow.Record, cols []int) (*array.Int64, error) {
-	if len(cols) > 1 {
-		return nil, fmt.Errorf("sorting by multiple columns isn't implemented yet")
+func (d Direction) comparison() int {
+	switch d {
+	case Ascending:
+		return -1
+	case Descending:
+		return 1
+	default:
+		return 0
 	}
-	indicesBuilder := array.NewInt64Builder(mem)
+}
 
+// SortingColumn describes a sorting column on a arrow.Record.
+type SortingColumn struct {
+	Index      int
+	Direction  Direction
+	NullsFirst bool
+}
+
+// SortRecord sorts given arrow.Record by columns.
+// Returns  *array.Int32 of  indices to sorted rows or record r.
+//
+// Comparison is made sequentially by each column. When rows are equal in the
+// first column we compare the rows om the second column and so on and so forth
+// until rows that are not equal are found.
+func SortRecord(
+	indicesBuilder *builder.OptInt32Builder,
```

Comment:
I don't like the API of having to instantiate a builder and then passing it in to `SortRecord`. Any reason you did it this way? Can we remove and instantiate within the `SortRecord` function?

Reply:
done 18f38f1